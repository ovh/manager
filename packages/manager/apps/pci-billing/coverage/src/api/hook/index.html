
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/api/hook</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> src/api/hook</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">8.97% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>58/646</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">55.55% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>5/9</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">80% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>4/5</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">8.97% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>58/646</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="useConsumption.tsx"><a href="useConsumption.tsx.html">useConsumption.tsx</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="586" class="abs low">0/586</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="586" class="abs low">0/586</td>
	</tr>

<tr>
	<td class="file high" data-value="useHistory.ts"><a href="useHistory.ts.html">useHistory.ts</a></td>
	<td data-value="96.66" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96.66" class="pct high">96.66%</td>
	<td data-value="60" class="abs high">58/60</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="8" class="abs medium">5/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="96.66" class="pct high">96.66%</td>
	<td data-value="60" class="abs high">58/60</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-12-09T10:51:49.479Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    