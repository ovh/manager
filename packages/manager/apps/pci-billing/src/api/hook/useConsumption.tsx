import { useMutation, useQuery } from '@tanstack/react-query';
import { COLD_ARCHIVE_GRID_DATA } from '@/constants';
import queryClient from '@/queryClient';
import {
  activateMonthlyBilling,
  getCurrentUsage,
  TCurrentUsage,
  TQuantity,
} from '../data/consumption';

const roundPrice = (num: number, fractionDigits = 2) =>
  Number(num.toFixed(fractionDigits));

const sumPrices = <T,>(items: T[], key: string) =>
  roundPrice(items?.reduce((sum, item) => sum + item[key], 0));

const initMonthlyInstanceList = (data: TCurrentUsage) => {
  if (!data?.monthlyUsage) {
    return {
      monthlyInstanceList: [],
      monthlyInstanceTotalPrice: 0,
    };
  }

  const monthlyInstanceList = data.monthlyUsage.instance.flatMap((instance) =>
    instance.details.map((detail) => ({
      ...detail,
      totalPrice: roundPrice(detail.totalPrice),
      reference: instance.reference,
      region: instance.region,
    })),
  ) as TInstance[];

  const monthlyInstanceTotalPrice = roundPrice(
    data.monthlyUsage.instance.reduce(
      (sum, instance) => sum + roundPrice(instance.totalPrice),
      0,
    ),
  );

  return {
    monthlyInstanceList,
    monthlyInstanceTotalPrice,
  };
};

const initHourlyInstanceList = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return { hourlyInstanceList: [], hourlyInstanceTotalPrice: 0 };
  }

  const hourlyInstanceList = data.hourlyUsage.instance.flatMap((instance) =>
    instance.details.map((detail) => ({
      ...detail,
      totalPrice: roundPrice(detail.totalPrice),
      reference: instance.reference,
      region: instance.region,
    })),
  ) as TInstance[];

  const hourlyInstanceTotalPrice = roundPrice(
    data.hourlyUsage.instance.reduce(
      (sum, instance) => sum + roundPrice(instance.totalPrice),
      0,
    ),
  );

  return {
    hourlyInstanceList,
    hourlyInstanceTotalPrice,
  };
};

const initObjectStorageList = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return {
      objectStorageList: [],
      objectStorageTotalPrice: 0,
    };
  }

  const objectStorageList = data.hourlyUsage.storage?.filter(
    (storage) => storage.type !== 'pca',
  );

  const objectStorageTotalPrice = roundPrice(
    objectStorageList.reduce(
      (sum, storage) => sum + roundPrice(storage.totalPrice),
      0,
    ),
  );

  return {
    objectStorageList,
    objectStorageTotalPrice,
  };
};

const initArchiveStorageList = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return {
      archiveStorageList: [],
      archiveStorageTotalPrice: 0,
    };
  }

  const archiveStorageList = data.hourlyUsage.storage?.filter(
    (storage) => storage.type === 'pca',
  );

  const archiveStorageTotalPrice = roundPrice(
    archiveStorageList.reduce(
      (sum, storage) => sum + roundPrice(storage.totalPrice),
      0,
    ),
  );

  return {
    archiveStorageList,
    archiveStorageTotalPrice,
  };
};

const initSnapshotList = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return {
      snapshotList: [],
      snapshotsTotalPrice: 0,
    };
  }

  const snapshotList = data.hourlyUsage.snapshot.map((snapshot) => ({
    ...snapshot,
    totalPrice: roundPrice(snapshot.totalPrice),
  }));

  const snapshotsTotalPrice = roundPrice(
    data.hourlyUsage.snapshot.reduce(
      (sum, snapshot) => sum + roundPrice(snapshot.totalPrice),
      0,
    ),
  );

  return {
    snapshotList,
    snapshotsTotalPrice,
  };
};

const initVolumeList = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return {
      volumeList: [],
      volumesTotalPrice: 0,
    };
  }

  const volumeList = data.hourlyUsage.volume.flatMap((volume) =>
    volume.details.map((detail) => ({
      ...detail,
      totalPrice: roundPrice(detail.totalPrice),
      type: volume.type,
      region: volume.region,
    })),
  );

  const volumesTotalPrice = roundPrice(
    data.hourlyUsage.volume.reduce(
      (sum, volume) => sum + roundPrice(volume.totalPrice),
      0,
    ),
  );

  return {
    volumeList,
    volumesTotalPrice,
  };
};

const initInstanceBandwidth = (data: TCurrentUsage) => {
  if (!data?.hourlyUsage) {
    return {
      bandwidthList: [],
      bandwidthTotalPrice: 0,
    };
  }

  const bandwidthList = data.hourlyUsage.instanceBandwidth.map(
    (bandwidthByRegion) => ({
      ...bandwidthByRegion,
      outgoingBandwidth: {
        ...bandwidthByRegion.outgoingBandwidth,
        totalPrice: roundPrice(bandwidthByRegion.outgoingBandwidth.totalPrice),
        quantity: {
          ...bandwidthByRegion.outgoingBandwidth.quantity,
          value: roundPrice(
            bandwidthByRegion.outgoingBandwidth.quantity.value || 0,
          ),
        },
      },
    }),
  );

  const bandwidthTotalPrice = roundPrice(
    bandwidthList.reduce(
      (sum, item) => sum + item.outgoingBandwidth.totalPrice,
      0,
    ),
  );

  return {
    bandwidthList,
    bandwidthTotalPrice,
  };
};

const sumColdArchiveTotalPriceByName = (
  coldArchives: TConsumptionDetail['coldArchive'],
  name: string,
) =>
  coldArchives.reduce((sum, object) => {
    if (object.name === name) {
      return sum + object?.totalPrice;
    }
    return sum;
  }, 0);

const sumColdArchiveBillingQuantitiesByName = (
  coldArchives: TConsumptionDetail['coldArchive'],
  name: string,
) =>
  coldArchives.reduce((sum, object) => {
    if (object.name === name) {
      return sum + object?.quantity?.value;
    }
    return sum;
  }, 0);

const reduceColdArchiveBillingInfo = (
  coldArchives: TConsumptionDetail['coldArchive'],
) =>
  Object.keys(COLD_ARCHIVE_GRID_DATA.FEE_TYPES).map((type) => {
    const name = type.toLowerCase();
    return {
      name,
      quantity: {
        value: sumColdArchiveBillingQuantitiesByName(coldArchives, name),
        unit:
          name === COLD_ARCHIVE_GRID_DATA.FEE_TYPES.RESTORE
            ? COLD_ARCHIVE_GRID_DATA.QUANTITY.UNIT
            : COLD_ARCHIVE_GRID_DATA.QUANTITY.HOUR,
      },
      totalPrice: sumColdArchiveTotalPriceByName(coldArchives, name),
      region: COLD_ARCHIVE_GRID_DATA.REGION,
    };
  });

const initResourceUsage = (data: TCurrentUsage, resourceType: string) => {
  let resources = (data?.resourcesUsage || [])
    .filter((resource) => resource.type === resourceType)
    .flatMap((resource) => resource.resources || [])
    .flatMap((resourceDetail) =>
      (resourceDetail.components || []).map((resourceComponent) => ({
        ...resourceComponent,
        region: resourceDetail.region,
      })),
    );

  if (resourceType === 'coldarchive') {
    resources = reduceColdArchiveBillingInfo(resources);
  }

  const totalPrice = sumPrices(resources, 'totalPrice');

  return {
    resources: resources || [],
    totalPrice: totalPrice || 0,
  };
};

export type TResourceUsage = {
  name: string;
  quantity: TQuantity;
  region: string;
  totalPrice: number;
};

export type TInstance = {
  instanceId: string;
  quantity: TQuantity;
  reference: string;
  region: string;
  totalPrice: number;
  activation: string;
};

export type TVolume = {
  quantity: TQuantity;
  region: string;
  totalPrice: number;
  type: string;
  volumeId: string;
};

export type TInstanceBandWith = {
  incomingBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  outgoingBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  region: string;
  totalPrice: number;
};

export type TStorage = {
  bucketName: string;
  incomingBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  incomingInternalBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  outgoingBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  outgoingInternalBandwidth: {
    quantity: TQuantity;
    totalPrice: number;
  };
  region: string;
  stored: {
    quantity: TQuantity;
    totalPrice: number;
  };
  totalPrice: number;
  type: string;
};

export type TSnapshot = {
  instance: {
    quantity: TQuantity;
    totalPrice: 0;
  };
  region: string;
  totalPrice: number;
  volume: string | null;
};

export type TConsumptionDetail = {
  hourlyInstances: TInstance[];
  monthlyInstances: TInstance[];
  objectStorages: TStorage[];
  archiveStorages: TStorage[];
  snapshots: TSnapshot[];
  volumes: TVolume[];
  bandwidthByRegions: TInstanceBandWith[];
  privateRegistry: TResourceUsage[];
  kubernetesLoadBalancer: TResourceUsage[];
  training: TResourceUsage[];
  notebooks: TResourceUsage[];
  aiDeploy: TResourceUsage[];
  coldArchive: TResourceUsage[];
  dataProcessing: TResourceUsage[];
  databases: TResourceUsage[];
  floatingIP: TResourceUsage[];
  publicIP: TResourceUsage[];
  gateway: TResourceUsage[];
  octaviaLoadBalancer: TResourceUsage[];
  totals: {
    total: number;
    hourly: {
      total: number;
      instance: number;
      objectStorage: number;
      archiveStorage: number;
      snapshot: number;
      volume: number;
      bandwidth: number;
      privateRegistry: number;
      kubernetesLoadBalancer: number;
      notebooks: number;
      coldArchive: number;
      serving: number;
      training: number;
      aiDeploy: number;
      dataProcessing: number;
      databases: number;
      floatingIP: number;
      gateway: number;
      octaviaLoadBalancer: number;
      publicIP: number;
    };
    monthly: {
      total: number;
      instance: number;
    };
  };
};

export const initializeTConsumptionDetail = (): TConsumptionDetail => ({
  hourlyInstances: [],
  monthlyInstances: [],
  objectStorages: [],
  archiveStorages: [],
  snapshots: [],
  volumes: [],
  bandwidthByRegions: [],
  privateRegistry: [],
  kubernetesLoadBalancer: [],
  training: [],
  notebooks: [],
  aiDeploy: [],
  coldArchive: [],
  dataProcessing: [],
  databases: [],
  floatingIP: [],
  publicIP: [],
  gateway: [],
  octaviaLoadBalancer: [],
  totals: {
    total: 0,
    hourly: {
      total: 0,
      instance: 0,
      objectStorage: 0,
      archiveStorage: 0,
      snapshot: 0,
      volume: 0,
      bandwidth: 0,
      privateRegistry: 0,
      kubernetesLoadBalancer: 0,
      notebooks: 0,
      coldArchive: 0,
      serving: 0,
      training: 0,
      aiDeploy: 0,
      dataProcessing: 0,
      databases: 0,
      floatingIP: 0,
      gateway: 0,
      octaviaLoadBalancer: 0,
      publicIP: 0,
    },
    monthly: {
      total: 0,
      instance: 0,
    },
  },
});

export const getConsumptionDetails = (usage: TCurrentUsage) => {
  const resourceMap = [
    { type: 'registry', key: 'privateRegistry' },
    { type: 'loadbalancer', key: 'kubernetesLoadBalancer' },
    { type: 'ai-notebook', key: 'notebooks' },
    { type: 'ai-serving-engine', key: 'serving' },
    { type: 'ai-training', key: 'training' },
    { type: 'data-processing-job', key: 'dataProcessing' },
    { type: 'databases', key: 'databases' },
    { type: 'coldarchive', key: 'coldArchive' },
    { type: 'floatingip', key: 'floatingIP' },
    { type: 'gateway', key: 'gateway' },
    { type: 'octavia-loadbalancer', key: 'octaviaLoadBalancer' },
    { type: 'ai-app', key: 'aiDeploy' },
    { type: 'publicip', key: 'publicIP' },
  ];

  const { resources, totals: hourlyTotals } = resourceMap.reduce(
    (acc, { type, key }) => {
      const { resources: result, totalPrice } = initResourceUsage(usage, type);
      acc.resources[key] = result;
      acc.totals[key] = totalPrice;
      return acc;
    },
    { resources: {}, totals: {} } as {
      resources: Record<string, TResourceUsage[]>;
      totals: Record<string, number>;
    },
  );

  const {
    monthlyInstanceList,
    monthlyInstanceTotalPrice,
  } = initMonthlyInstanceList(usage);

  const {
    hourlyInstanceList,
    hourlyInstanceTotalPrice,
  } = initHourlyInstanceList(usage);

  const { objectStorageList, objectStorageTotalPrice } = initObjectStorageList(
    usage,
  );
  const {
    archiveStorageList,
    archiveStorageTotalPrice,
  } = initArchiveStorageList(usage);
  const { snapshotList, snapshotsTotalPrice } = initSnapshotList(usage);
  const { volumeList, volumesTotalPrice } = initVolumeList(usage);
  const { bandwidthList, bandwidthTotalPrice } = initInstanceBandwidth(usage);

  const totals = {
    hourly: {
      total: 0,
      ...hourlyTotals,
      instance: hourlyInstanceTotalPrice,
      objectStorage: objectStorageTotalPrice,
      archiveStorage: archiveStorageTotalPrice,
      snapshot: snapshotsTotalPrice,
      volume: volumesTotalPrice,
      bandwidth: bandwidthTotalPrice,
    },
    monthly: {
      total: monthlyInstanceTotalPrice,
      instance: monthlyInstanceTotalPrice,
    },
  };

  totals.hourly.total = roundPrice(
    Object.values(totals.hourly).reduce((sum, price) => sum + price, 0),
  );

  const finalTotal = roundPrice(totals.hourly.total + totals.monthly.total);

  return {
    hourlyInstances: hourlyInstanceList,
    monthlyInstances: monthlyInstanceList,
    objectStorages: objectStorageList,
    archiveStorages: archiveStorageList,
    snapshots: snapshotList,
    volumes: volumeList,
    bandwidthByRegions: bandwidthList,
    ...resources,
    totals: {
      ...totals,
      total: finalTotal,
    },
  } as TConsumptionDetail;
};

export const useGeTCurrentUsage = (projectId: string) =>
  useQuery({
    queryKey: [projectId, 'current', 'usage'],
    queryFn: async () => {
      const currentUsage = await getCurrentUsage(projectId);
      return currentUsage;
    },
    select: (data) => getConsumptionDetails(data),
  });

type ActivateMonthlyBillingProps = {
  projectId: string;
  onError: (error: Error) => void;
  onSuccess: () => void;
};

export const useActivateMonthlyBilling = ({
  projectId,
  onError,
  onSuccess,
}: ActivateMonthlyBillingProps) => {
  const mutation = useMutation({
    mutationFn: async (instanceId: string) =>
      activateMonthlyBilling(projectId, instanceId),
    onError,
    onSuccess: async () => {
      await queryClient.invalidateQueries({
        queryKey: [projectId, 'current', 'usage'],
      });

      onSuccess();
    },
  });

  return {
    activateMonthlyBilling: (instanceId: string) => mutation.mutate(instanceId),
    ...mutation,
  };
};
