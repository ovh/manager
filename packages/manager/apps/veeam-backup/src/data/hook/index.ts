export * from './useOrganizations';
export * from './useVeeamBackup';
export * from './useOrganizations';
