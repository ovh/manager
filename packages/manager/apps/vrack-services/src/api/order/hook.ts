import { useQuery, useQueryClient } from '@tanstack/react-query';
import {
  getOrderStatus,
  getOrderDetails,
  getOrderDetailsList,
  getOrderList,
  getOrderData,
} from './services';
import { OrderDetail, OrderStatus, OrderData } from './order.type';

export type DetailedOrder = OrderData & {
  status: OrderStatus;
  details: OrderDetail[];
};

const toDetailedOrder = async (orderId: number): Promise<DetailedOrder> => {
  const orderData = await getOrderData(orderId);
  const orderProductList = await getOrderDetailsList(orderId);
  const detailsResponse = await Promise.all(
    orderProductList.data.map((detailId) =>
      getOrderDetails({ orderId, detailId }),
    ),
  );
  const statusResponse = await getOrderStatus(orderId);
  return {
    orderId,
    ...orderData.data,
    status: statusResponse.data,
    details: detailsResponse.map((detail) => detail.data),
  };
};

const getOrderIdListWithDescription = (
  detailedOrderList: DetailedOrder[] = [],
  orderDetailDescription: string,
) =>
  detailedOrderList.filter(({ details }) =>
    details.some(({ description }) =>
      description.includes(orderDetailDescription),
    ),
  );

export enum OrderDescription {
  vrack = 'vRack',
  vrackServices = 'vRack Services',
}

export const getDeliveringOrderQueryKey = (description: OrderDescription) => [
  `deliveringOrders-${description}`,
];

export const getDeliveringOrderList = (
  orderDetailDescription: string,
) => async () => {
  const todayMinus24Hours = new Date();
  todayMinus24Hours.setTime(Date.now() - 24 * 3600 * 1000);
  const orderList = await getOrderList({ dateFrom: todayMinus24Hours });

  const orderIdsWithStatus = await Promise.all(
    orderList.data.map(async (orderId) => {
      const status = await getOrderStatus(orderId);
      return { orderId, status: status.data };
    }),
  );
  const deliveringOrderList = orderIdsWithStatus
    .filter(({ status }) => status === 'delivering')
    .map(({ orderId }) => orderId);

  const detailedOrderList = await Promise.all(
    deliveringOrderList.map(toDetailedOrder),
  );

  return getOrderIdListWithDescription(
    detailedOrderList,
    orderDetailDescription,
  );
};

export const getPollingOderStatusQueryKey = (pollingKey: OrderDescription) => [
  `/pollOrderStatus-${pollingKey}`,
];

/**
 * Check for orders that contain a certain type of product and refresh the order untill it's delivered
 */
export const useOrderPollingStatus = ({
  pollingKey,
  orderList = [],
  queryToInvalidateOnDelivered,
  pollingInterval = 120000,
}: {
  pollingKey: OrderDescription;
  orderList: DetailedOrder[];
  queryToInvalidateOnDelivered: string[];
  pollingInterval?: number;
}) => {
  const queryClient = useQueryClient();
  return useQuery({
    queryKey: getPollingOderStatusQueryKey(pollingKey),
    queryFn: async () => {
      if (orderList.every(({ status }) => status === 'delivered')) {
        queryClient.invalidateQueries({
          queryKey: queryToInvalidateOnDelivered,
        });
      }
      return orderList.map(({ status }) => status);
    },
    refetchInterval: (data) => {
      if (!data || data.length === 0) {
        return 0;
      }
      return data.every((status) => status === 'delivered')
        ? pollingInterval
        : 0;
    },
  });
};
