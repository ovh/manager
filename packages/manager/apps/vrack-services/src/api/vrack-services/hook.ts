import React from 'react';
import { useQuery, useQueryClient } from '@tanstack/react-query';
import { getListingIcebergQueryKey, getTaskList } from './get';
import { Status } from '../api.type';

export type UseTaskPollingParams = {
  vrackServicesId: string;
  initialRefetchInterval?: number;
  pollingInterval?: number;
};

/**
 * Check for orders that contain a certain type of product and refresh the order untill it's delivered
 */
export const useTaskPolling = ({
  vrackServicesId,
  initialRefetchInterval = 0,
  pollingInterval = 30000,
}: UseTaskPollingParams) => {
  const queryClient = useQueryClient();
  const [refetchInterval, setRefetchInterval] = React.useState(
    initialRefetchInterval,
  );

  const query = useQuery({
    queryKey: ['taskPolling', vrackServicesId],
    queryFn: async () => {
      const response = await getTaskList({ vrackServicesId });
      if (response.data.every(({ status }) => status === Status.DONE)) {
        setRefetchInterval(0);
        queryClient.invalidateQueries({
          queryKey: getListingIcebergQueryKey,
        });
      }
      return response.data.filter(({ status }) => status !== Status.DONE);
    },
    refetchInterval,
  });

  React.useEffect(() => {
    if (query.data?.length > 0) {
      setRefetchInterval(pollingInterval);
    }
  }, [query.data?.length]);

  return query;
};
