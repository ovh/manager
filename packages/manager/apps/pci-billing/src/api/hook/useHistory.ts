import { useQuery } from '@tanstack/react-query';
import { isSameMonth } from 'date-fns';
import { useComputeDate } from '@/components/history/useComputeDate.hook';
import { getCurrentUsage } from '../data/consumption';
import {
  getUsageHistory,
  getUsageHistoryPeriod,
  TUsageHistoryPeriod,
} from '../data/history';
import { getConsumptionDetails } from './useConsumption';

export const useUsageHistoryPeriod = (
  projectId: string,
  from: string,
  to: string,
) =>
  useQuery({
    queryKey: [projectId, from, to, 'history', 'usage'],
    queryFn: async () => {
      const historyPeriod = await getUsageHistoryPeriod(projectId, from, to);
      return historyPeriod?.length
        ? historyPeriod[0]
        : ({} as TUsageHistoryPeriod);
    },
  });

export const useGetUsageHistory = (
  projectId: string,
  periodDetail: TUsageHistoryPeriod,
) => {
  const { billingDate } = useComputeDate();

  return useQuery({
    queryKey: [projectId, periodDetail?.id, 'history', 'usage'],
    queryFn: async () => {
      const historyUsage = await getUsageHistory(projectId, periodDetail?.id);
      let monthlyDetails;

      if (isSameMonth(new Date(), billingDate)) {
        const currentUsage = await getCurrentUsage(projectId);
        monthlyDetails = currentUsage?.monthlyUsage;
      } else if (
        new Date(periodDetail.period.to).getUTCMonth() ===
        billingDate.getMonth()
      ) {
        monthlyDetails = historyUsage?.monthlyUsage;
      }

      return { ...historyUsage, monthlyUsage: monthlyDetails };
    },
    enabled: !!periodDetail?.id,
    select: (data) => getConsumptionDetails(data),
  });
};
