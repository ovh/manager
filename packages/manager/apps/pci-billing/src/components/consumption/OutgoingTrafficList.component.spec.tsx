import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { TInstanceBandWith } from '@/api/hook/useConsumption';
import OutgoingTrafficList from './OutgoingTrafficList.component';

describe('OutgoingTrafficList', () => {
  const mockInstanceBandwidths = [
    {
      incomingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 0,
        },
        totalPrice: 0,
      },
      outgoingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 7.33,
        },
        totalPrice: 0,
      },
      region: 'SGP1',
      totalPrice: 0,
    },
    {
      incomingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 0,
        },
        totalPrice: 0,
      },
      outgoingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 1,
        },
        totalPrice: 0,
      },
      region: 'AP-SOUTH-MUM-1',
      totalPrice: 0,
    },
    {
      incomingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 0,
        },
        totalPrice: 0,
      },
      outgoingBandwidth: {
        quantity: {
          unit: 'GiB',
          value: 4.09,
        },
        totalPrice: 0,
      },
      region: 'SYD1',
      totalPrice: 0,
    },
  ] as TInstanceBandWith[];

  it('matches snapshot with existing storages', () => {
    const { asFragment } = render(
      <OutgoingTrafficList instanceBandwidths={mockInstanceBandwidths} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with empty storages', () => {
    const { asFragment } = render(
      <OutgoingTrafficList instanceBandwidths={[]} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
