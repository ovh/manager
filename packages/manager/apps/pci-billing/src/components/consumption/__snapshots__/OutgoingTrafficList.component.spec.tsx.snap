// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OutgoingTrafficList > matches snapshot with empty storages 1`] = `
<DocumentFragment>
  <div
    class="my-3"
  >
    <osds-text
      class="mb-5"
      color="text"
      level="body"
      size="400"
    >
      cpbc_no_consumption_data
    </osds-text>
  </div>
</DocumentFragment>
`;

exports[`OutgoingTrafficList > matches snapshot with existing storages 1`] = `
<DocumentFragment>
  <div
    class="my-3"
  >
    <div>
      <div
        class="contents px-[1px] "
      >
        <osds-table
          class="overflow-x-visible"
        >
          <!---->
          <table
            class="w-full border-collapse"
          >
            <thead>
              <tr>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-region"
                  >
                    <span>
                      <osds-text
                        color="primary"
                        hue="800"
                        level="body"
                        size="500"
                      >
                        cpbc_hourly_instance_trafic_region
                      </osds-text>
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <osds-icon
                        class="invisible"
                        color="primary"
                        name="arrow-down"
                        size="xxs"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-traffic"
                  >
                    <span>
                      <osds-text
                        color="primary"
                        hue="800"
                        level="body"
                        size="500"
                      >
                        cpbc_hourly_instance_trafic_trafic
                      </osds-text>
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <osds-icon
                        class="invisible"
                        color="primary"
                        name="arrow-down"
                        size="xxs"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-price"
                  >
                    <span>
                      <osds-text
                        color="primary"
                        hue="800"
                        level="body"
                        size="500"
                      >
                        cpbc_hourly_instance_trafic_price
                      </osds-text>
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <osds-icon
                        class="invisible"
                        color="primary"
                        name="arrow-down"
                        size="xxs"
                      />
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[var(--ods-color-blue-200)]"
              >
                <td
                  class="text-left pl-4"
                >
                  <osds-text
                    color="text"
                    level="body"
                  >
                    manager_components_region_SGP_micro
                  </osds-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      7.87/1024 unit_size_GB
                    </osds-text>
                  </div>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      0 undefined
                    </osds-text>
                  </div>
                </td>
              </tr>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[var(--ods-color-blue-200)]"
              >
                <td
                  class="text-left pl-4"
                >
                  <osds-text
                    color="text"
                    level="body"
                  >
                    manager_components_region_MUM_micro
                  </osds-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      cpbc_hourly_instance_trafic_unlimitted
                    </osds-text>
                  </div>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      cpbc_hourly_instance_trafic_included
                    </osds-text>
                  </div>
                </td>
              </tr>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[var(--ods-color-blue-200)]"
              >
                <td
                  class="text-left pl-4"
                >
                  <osds-text
                    color="text"
                    level="body"
                  >
                    manager_components_region_SYD_micro
                  </osds-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      4.39/1024 unit_size_GB
                    </osds-text>
                  </div>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div>
                    <osds-text
                      color="text"
                      level="body"
                    >
                      0 undefined
                    </osds-text>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </osds-table>
      </div>
      <osds-pagination
        class="flex xs:justify-start md:justify-end"
        default-current-page="1"
        default-items-per-page="10"
        total-items="3"
        total-pages="1"
      >
        <span
          class="mr-3"
          slot="before-total-items"
        >
          common_pagination_of
        </span>
        <span
          class="ml-3"
          slot="after-total-items"
        >
          common_pagination_results
        </span>
      </osds-pagination>
    </div>
  </div>
</DocumentFragment>
`;
